### PR TITLE
listener manager: handle listener creation failures on workers

### PR DESCRIPTION
**Pull request by mattklein123**
This commit properly handles listener creation failures on worker threads.
Specifically, the race condition that can happen between bind() and
listen(). (To be perfectly honest I don't understand how this can happen
at the OS level but we have seen it in practice).

This turned into somewhat of a complicated change. It does the following
things:
1) Add an add listener completion which can tell the listener manager
   if a creation failed. As part of this work, locking was replaced with
   posts inside the listener manager which makes synchronization simpler.
2) The failed listener is removed, in the hope that a future addition
   it might succeed at that point.
3) Listeners are now always added to workers via post.
4) Because of (3), this caused race conditions in the integration tests.
   The integration startup code needed to be refactored so that we now
   wait for all listeners to be up before starting to run the tests.
5) Because of (4) it exposed an issue in the TCP proxy integration test
   with trying to handle a response from the client SSL auth filter which
   happens before the listeners are actually up. To fix this, I removed
   the code in the tests trying to handle client SSL auth filter calls.
   I also replaced all "random" upstream clusters with a host pointing
   at reserved port 4 which should always fail in any reasonable case.
   This still has value in testing startup with failure, but won't block
   startup. If we want to bring back testing of initial startup things
   like SDS, RDS, etc. we will need to revisit how this works but IMO
   this is fine for now.

### CODE REVIEW

**Comment by mattklein123** [comment]
@lyft/network-team @htuch @dnoe @alyssawilk 

@alyssawilk this is going to conflict with your flow control change in the tests. Also LMK what you think about that part of the change. We can get your change in first and then I can merge master into this.

File: test/integration/tcp_proxy_integration_test.cc [hunk 1]
```diff
@@ -84,15 +72,6 @@ TEST_P(TcpProxyIntegrationTest, SendTlsToTlsListener) {
...
       },
-      // Set up the initial REST response for the ssl_auth filter to avoid the async client doing
-      // reconnects.  Loopback is whitelisted by default so no response payload is necessary.
-      [&]() -> void {
-        fake_rest_connection = fake_upstreams_[1]->waitForHttpConnection(*dispatcher_);
-        request = fake_rest_connection->waitForNewStream();
-        request->waitForEndStream(*dispatcher_);
-        request->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}}, false);
-        request->encodeData(0, true);
-      },
```
Comment (alyssawilk) [inline test/integration/tcp_proxy_integration_test.cc]: Is this the conflict you meant?
If so please feel free to check in ASAP, because it is way cleaner and I'm happy to do the merges.

**Comment by alyssawilk** [comment]
FWIW 1k is sadly nowhere enough - I've spent more of the day debugging a
1:20k failure which I bet this is the fix to.

(I got as far as figuring the SSL auth code was getting a bad FD which
aligns well with the idea not everything is set up right)

File: test/integration/tcp_proxy_integration_test.cc (hunk 1 above)
Comment (mattklein123) [inline test/integration/tcp_proxy_integration_test.cc]: Sure either way. We can take this through review and merge first if you want. I agree that as long as you don't have a problem with the change it does make the tests substantially easier to grok.

**Comment by mattklein123** [comment]
> Asking as you seem to have uncovered a bunch of concurrency issues in the integration tests, just wanted to check there aren't any more.

Not exactly, it's really an ordering issue. The old code would add listeners on the startup thread. This no longer happens, so new sync is needed. This also broke the TCP proxy tests. However I will run a long test cycle locally to make sure it works with multiple iterations. @alyssawilk I have no doubt there are still some edge cases in the existing code.

File: source/server/listener_manager_impl.cc [hunk 2]
```diff
@@ -379,11 +374,33 @@ std::vector<std::reference_wrapper<Listener>> ListenerManagerImpl::listeners() {
...
+    // avoid locking.
+    server_.dispatcher().post([this, success, &listener]() -> void {
+      // It is theoretically possible for a listener to get added on 1 worker but not the others.
+      // The following code will yield 1 critical log and 1 stat. It will also cause listener
+      // removal a single time. Note also that that drain/removal can race with addition. It's
+      // guaranteed that workers process remove after add so this should be fine.
+      if (!success && !listener.onListenerCreateFailure()) {
+        // TODO(mattklein123): In addition to a critical log and a stat, we should consider adding
+        //                     a startup option here to cause the server to exit. I think we
+        //                     probably want this at Lyft but I will do it in a follow up.
```
Comment (htuch) [inline source/server/listener_manager_impl.cc]: I think we would also want to fail a dry run canary here too.

File: source/server/listener_manager_impl.h [hunk 3]
```diff
@@ -173,6 +171,16 @@ class ListenerImpl : public Listener,
                bool workers_started, uint64_t hash);
   ~ListenerImpl();
 
+  /**
+   * Called when a listener failed to be actually created on a worker.
+   * @return TRUE if this is not the first create failure (via multiple workers).
```
Comment (htuch) [inline source/server/listener_manager_impl.h]: @return TRUE if we have seen more than one worker failure

File: test/config/integration/server.json [hunk 4]
```diff
@@ -254,7 +254,7 @@
         "connect_timeout_ms": 5000,
         "type": "static",
         "lb_type": "round_robin",
-        "hosts": [{"url": "tcp://{{ ip_loopback_address }}:12000"}]
+        "hosts": [{"url": "tcp://{{ ip_loopback_address }}:4"}]
```
Comment (htuch) [inline test/config/integration/server.json]: Why this change? Are we making meaningful use of these?

File: test/integration/server.cc [hunk 5]
```diff
@@ -64,6 +68,24 @@ IntegrationTestServer::~IntegrationTestServer() {
   thread_->join();
 }
 
+void IntegrationTestServer::onWorkerListenerAdded() {
+  if (on_worker_listener_added_cb_) {
+    on_worker_listener_added_cb_();
+  }
+
+  std::unique_lock<std::mutex> gaurd(listeners_mutex_);
```
Comment (htuch) [inline test/integration/server.cc]: Nit: s/gaurd/guard/ (or is it gourd? :) )

File: source/server/listener_manager_impl.cc [hunk 6]
```diff
@@ -379,11 +374,33 @@ std::vector<std::reference_wrapper<Listener>> ListenerManagerImpl::listeners() {
   return ret;
 }
 
+void ListenerManagerImpl::addListenerToWorker(Worker& worker, ListenerImpl& listener) {
+  worker.addListener(listener, [this, &listener](bool success) -> void {
+    // The add listener completion runs on the worker thread. Post back to the main thread to
+    // avoid locking.
+    server_.dispatcher().post([this, success, &listener]() -> void {
+      // It is theoretically possible for a listener to get added on 1 worker but not the others.
+      // The following code will yield 1 critical log and 1 stat. It will also cause listener
```
Comment (htuch) [inline source/server/listener_manager_impl.cc]: I would rephrase this as it kind of confused me on the first pass, something to the effect of "The below check with onListenerCreateFailure() is there to ensure we execute the removal/logging/stats at most once on failure."

File: test/config/integration/server.json (hunk 4 above)
Comment (mattklein123) [inline test/config/integration/server.json]: There is no great way to put a comment about this, but it does do more than nothing. Basically it verifies real flow of connecting to various things and then immediately failing. So it verifies that init flow doesn't get blocked, no crashes, etc. I admit it's not perfect but I don't think it's worth deleting. LMK what you think and if you want me to document somewhere.

File: source/server/listener_manager_impl.cc (hunk 2 above)
Comment (mattklein123) [inline source/server/listener_manager_impl.cc]: I don't think we would actually add a listener all the way to this point in a dry run situation? It's not completely clear to me what happens in a dry run (I saw your other comment). I think we should track this separately?

**Comment by mattklein123** [comment]
@htuch updated

File: test/config/integration/server.json (hunk 4 above)
Comment (htuch) [inline test/config/integration/server.json]: OK, so just to clarify, you changed these values to 4 here to ensure that they really would fail and not conflict with something on 12000? Might be useful to have this documented in some JSON comments..

Comment (mattklein123) [inline test/config/integration/server.json]: Yes. There are no JSON comments. :( Any other place I can put this?

File: test/test_common/utility.h [hunk 7]
```diff
@@ -95,7 +95,7 @@ class TestUtility {
 class ConditionalInitializer {
 public:
   /**
-   * Set the conditional to ready, should only be called once.
+   * Set the conditional to ready.
```
Comment (htuch) [inline test/test_common/utility.h]: Maybe explain that `waitReady()` resets, but in between `waitReady()`, `setReady()` should only be called once.

File: test/config/integration/server.json (hunk 4 above)
Comment (htuch) [inline test/config/integration/server.json]: Maybe don't worry then, I think anywhere else would just be confusing to the reader. I think this will be better in v2 - we won't be using manually crafted JSON configs for integration tests, but protobuf objects that we build dynamically for each different integration test. Then it can be in the code where we add these objects.

**Comment by mattklein123** [comment]
@htuch comment updated
